### PR TITLE
Doc endorsements

### PR DESCRIPTION
**Pull request by damip**
Explains how endorsements work

### CODE REVIEW

**Comment by damip** [comment]
@qdrn @leoloco Can you fill up the section "Choosing the value of `E`" ?

**Comment by damip** [comment]
@adrien-zinger Please check it out as well and give any feedback

**Comment by damip** [comment]
@jcbaillie FYI

**Review by jcbaillie — APPROVED** [review]

File: docs/technical-doc/Endorsements.rst [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+Massa uses the Proof-of-Stake selection mechanism with Nakamoto consensus.
+In that context, when there are multiple cliques in close competition,
+we want all nodes to converge towards a single clique as fast as possible to minimize finality time and maximize the quality of the consensus.
+To achieve this, we draw inspiration from Tezos and introduce the concept of Endorsement.
+
+Basic principle
+===============
+
+At every slot `S`, we use the existing Proof-of-Stake selection mechanism to not only draw the block creator for that slot,
+but also `E` other stakers that are responsible for creating Endorsements that can be included in the blocks of slot `S`.
```
Comment (adrien-zinger) [inline docs/technical-doc/Endorsements.rst:18]:
```suggestion
but also `E` other stakers that are responsible for creating Endorsements that can be included in the block of slot `S`.
```

File: docs/technical-doc/Endorsements.rst [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+Every endorsement that is selected to be included in blocks of slot `S` is signed by its creator `C`
```
Comment (adrien-zinger) [inline docs/technical-doc/Endorsements.rst:19]: I would split that sentence.

A block includes some endorsements. An endorsement contains the "endorsed" hash of the parent block. C, the creator of the endorsements, would choose that parent in the same thread of the slot 'S'. As well the endorsement's creators vote for the parent of the block in their own thread.

Every endorsement that is selected to be included in the Block of slot 'S' => Do you mean "Every endorsements in the block of slot S" ?

I guess the second sentence "Blocks in slot `S` can only include endorsements that endorse the block's parent block in the thread of `S`." is redundant, or can be resumed.

File: docs/technical-doc/Endorsements.rst [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+and contains the "endorsed" hash of the block that `C` would choose as the parent of a block at slot `S` in thread `S.thread`.
+Blocks in slot `S` can only include endorsements that endorse the block's parent block in the thread of `S`.
+In other words, endorsement producers vote for which parent should be chosen by blocks in their own thread.
+
+The likelihood of the attacker getting lucky and being selected for `N` consecutive PoS draws to attack/censor the system decays exponentially with `N`.
```
Comment (adrien-zinger) [inline docs/technical-doc/Endorsements.rst:24]:
```suggestion
The likelihood of the attacker getting lucky and being selected for `N` consecutive PoS draws to attack/censor the system decays exponentially.
```

File: docs/technical-doc/Endorsements.rst [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+.. code:: rust
+
+  pub struct Endorsement {
+    /// slot in which the endorsement can be included
+    pub slot: Slot,
+    /// endorsement index inside the including block
+    pub index: u32,
+    /// hash of endorsed block
+    pub endorsed_block: BlockId,
```
Comment (adrien-zinger) [inline docs/technical-doc/Endorsements.rst:45]: We can precise that the endorsed block is the selected parent in the thread of the slot. Selected by the producer of the endorsement.

File: docs/technical-doc/Endorsements.rst (hunk 1 above)
Comment (adrien-zinger) [reply docs/technical-doc/Endorsements.rst:18]: We can find only one block by slot right? :thinking:

**Comment by qdrn** [comment]
> @qdrn @leoloco Can you fill up the section "Choosing the value of `E`" ?

We are working on it :) Can't fill up for now. Do you want to merge now and we fill it up later ?

File: docs/technical-doc/Endorsements.rst (hunk 1 above)
Comment (damip) [reply docs/technical-doc/Endorsements.rst:18]: There can be multiple blocks at a given slot, but they would necessarily be from different cliques.

Example: I am selected to create a block a slot S but instead I create 2 different blocks at that slot.

**Comment by damip** [comment]
> > @qdrn @leoloco Can you fill up the section "Choosing the value of `E`" ?
> 
> We are working on it :) Can't fill up for now. Do you want to merge now and we fill it up later ?

ok

File: docs/technical-doc/Endorsements.rst (hunk 4 above)
Comment (damip) [reply docs/technical-doc/Endorsements.rst]: added

File: docs/technical-doc/Endorsements.rst (hunk 2 above)
Comment (damip) [reply docs/technical-doc/Endorsements.rst:19]: rewrote the paragrpah, hopefully more clearly

**Comment by damip** [comment]
All comments have been addressed. Merging

**Review by adrien-zinger — APPROVED** [review]

**Comment by damip** [comment]
bors merge

**Pull request merged**